rafthttp/stream.go

```diff
@@ -0,0 +1,353 @@
+		default:
+			log.Panicf("rafthttp: unhandled stream type %s", conn.t)
+		}
+		flusher = conn.Flusher
+		closer = conn.Closer
+		heartbeatc, msgc = tickc, cw.msgc
+		cw.working = true
+	}
+
+	for {
+		select {
+		case <-heartbeatc:
+			if err := enc.encode(linkHeartbeatMessage); err != nil {
+				log.Printf("rafthttp: encountered error heartbeating on stream: %v", err)
+				reset()
+				continue
+			}
+			flusher.Flush()
+		case m := <-msgc:
+			if t == streamTypeMsgApp && m.Term != msgAppTerm {
+				// TODO: reasonable retry logic
+				continue
+			}
+			if err := enc.encode(m); err != nil {
+				log.Printf("rafthttp: encountered error encoding to stream: %v", err)
+				reset()
+				continue
+			}
+			flusher.Flush()
+		case conn := <-cw.connc:
+			reset()
+			useConn(conn)
+		case <-cw.stopc:
+			reset()
+			close(cw.done)
+			return
+		}
+	}
+}
+
+func (cw *streamWriter) isWorking() bool {
+	cw.mu.Lock()
+	defer cw.mu.Unlock()
+	return cw.working
+}
+
+func (cw *streamWriter) attach(conn *outgoingConn) bool {
+	select {
+	case cw.connc <- conn:
+		return true
+	case <-cw.done:
+		return false
+	}
+}
+
+func (cw *streamWriter) stop() {
+	close(cw.stopc)
+	<-cw.done
+}
+
+// streamReader is a long-running worker that dials to the remote stream
+// endponit and reads messages from the response body returned.
+type streamReader struct {
+	tr       http.RoundTripper
+	u        string
+	t        streamType
+	from, to types.ID
+	cid      types.ID
+
+	mu         sync.Mutex
+	msgAppTerm uint64
+	req        *http.Request
+	closer     io.Closer
+
+	recvc chan<- raftpb.Message
+	dialc chan struct{}
+	stopc chan struct{}
+	done  chan struct{}
+}
+
+func startStreamReader(tr http.RoundTripper, u string, t streamType, from, to, cid types.ID, recvc chan<- raftpb.Message) *streamReader {
+	r := &streamReader{
+		tr:    tr,
+		u:     u,
+		t:     t,
+		from:  from,
+		to:    to,
+		cid:   cid,
+		recvc: recvc,
+		dialc: make(chan struct{}, 1),
+		stopc: make(chan struct{}),
+		done:  make(chan struct{}),
+	}
+	go r.run()
+	return r
+}
+
+func (cr *streamReader) run() {
+	var dec decoder
+	nextc := make(chan struct{}, 1)
+	sendNext := func() {
+		select {
+		case nextc <- struct{}{}:
+		default:
+		}
+	}
+	if cr.t != streamTypeMsgApp {
+		sendNext()
+	}
+	for {
+		select {
+		case <-cr.dialc:
+			cr.resetCloser()
+		case <-cr.stopc:
+			cr.resetCloser()
+			close(cr.done)
+			return
+		case <-nextc:
+		}
+		if !cr.isWorking() {
+			rc, err := cr.dial()
+			if err != nil {
+				log.Printf("rafthttp: roundtripping error: %v", err)
+				time.AfterFunc(100*time.Millisecond, sendNext)
+				continue
+			}
+			cr.mu.Lock()
+			switch cr.t {
+			case streamTypeMsgApp:
+				dec = &msgAppDecoder{r: rc, local: cr.from, remote: cr.to, term: cr.msgAppTerm}
+			case streamTypeGeneral:
+				dec = &messageDecoder{r: rc}
+			default:
+				log.Panicf("rafthttp: unhandled stream type %s", cr.t)
+			}
+			cr.closer = rc
+			cr.mu.Unlock()
+		}
+		m, err := dec.decode()
```

Comment:
can we put decode into another long running go-routine and it can greatly simplify the code logic.